### PR TITLE
Implemented review requests from pull request #44

### PR DESCRIPTION
**Pull request by Jochen-A-Fuerbacher**
Hi,
I implemented the review requests from pull request 44 ( https://github.com/jenkinsci/build-timeout-plugin/pull/44 ).

Jochen

### CODE REVIEW

**Comment by ikedam** [comment]
Note from #44 
* Difference from [Naginator plugin](https://wiki.jenkins-ci.org/display/JENKINS/Naginator+Plugin).
    * Naginator plugin requres builds failed to retrigger them.
    * That results failure notification mails are sent for every retrying.
    * This feature enables following configuration:
        * Retries builds silently.
        * Send a notification mail only when a build really fails.

**Comment by ikedam** [comment]
This indicates an option to fail the build if retries exceeded is useful.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            
+            if(build.getCause(BuildTimeoutAbortAndRestartCause.class) != null){
+                count++;
+            }
+         
+            if (count >= this.maxRestarts) {
+                return false;
+            }
+            build = build.getPreviousBuild();
+        }
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: I believe this is not what you want to do.
For example, if maxRestarts is 2:
* build1: aborted. triggers rtey.
* build2: retry of build1, aborted. triggers retry.
* build3: retry of build2, aborted. exceeded retrying.

In this case, new builds never trigger retries as there's already two retying builds in the history.
Using `BuildTimeoutAbortAndrestartCause#build` is recommended.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            ParametersAction action = build.getAction(ParametersAction.class);
+            build.getRootBuild().getProject().scheduleBuild(0, new BuildTimeoutAbortAndRestartCause(build), action);
+        }
+        return true;
+    }
+    
+    public class BuildTimeoutAbortAndRestartCause extends Cause {
+        
+
+        Run<?, ?> build;
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: You'd be interested in http://javadoc.jenkins-ci.org/jenkins/model/lazy/BuildReference.html
It might make sence to extend `UpstreamCause` as it should preserve relationships between original upstream builds.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, BuildListener listener, long effectiveTimeout) {
+        
+        new AbortOperation().perform(build, listener, effectiveTimeout);
+                
+        if (checkMaxRestarts(build)) {
+
+            ParametersAction action = build.getAction(ParametersAction.class);
+            build.getRootBuild().getProject().scheduleBuild(0, new BuildTimeoutAbortAndRestartCause(build), action);
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: Also inheriting `CauseAction` should be useful.
Please have a look at `NaginatorRetryAction#sceduleBuild`.
https://github.com/jenkinsci/naginator-plugin/blob/master/src/main/java/com/chikli/hudson/plugin/naginator/NaginatorRetryAction.java

**Comment by ikedam** [comment]
It makes sense that rescheduling the build only when a timeout occurs. That cannot be achieved with naginator-plugin.
However, At the same time, it 's preferred to integrate rescheduling features into one plugin.

What we really need is an operation to have naginator-plugin reschedule the build, isn't it?
That requires to extend naginator-plugin to accept  commandes to reschedule from other plugins and I'll work for that.
Would you let me have time for that work?

Continuing this pull request is useful as you can use this feature locally or naginator-plugin might not like that extension.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: +1. Causes should be copied. Otherwise there is a risk of security issues, because the rescheduled build may get authenticated differently in the Jenkins core (e.g. see Authorize Project)

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java (hunk 2 above)
Comment (Jochen-A-Fuerbacher) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: When we use `UpstreamCause`, then the default short description is shown. We do not want this. Overriding `getShortDescription()` doesn't work in this case.

**Comment by Jochen-A-Fuerbacher** [comment]
@ikedam Yes, that would be nice.

**Comment by ikedam** [comment]
I sent a pull request to naginator-plugin.
https://issues.jenkins-ci.org/browse/JENKINS-29715
https://github.com/jenkinsci/naginator-plugin/pull/16

**Comment by ikedam** [comment]
I found this feature requested long time ago: [JENKINS-8947](https://issues.jenkins-ci.org/browse/JENKINS-8947)

**Comment by ikedam** [comment]
Sorry to have kept you waiting.

Now [naginator-plugin-1.16alpha-1](http://repo.jenkins-ci.org/releases/org/jenkins-ci/plugins/naginator/1.16-alpha-1/) is availeble, which allows us trigger naginator-plugin from build-timeout plugin.
The outline to integrate:
* Add optional dependency to naginator in pom.xml
    * We use the alpha release only during developing. I believe 1.16 will be released before we finish this development. 
    * I don't think I have to upgrade the target Jenkins version even though naginator-plugin targets Jenkins 1.554 (build-timeout targets 1.466). Naginator-plugin doesn't depend on any features introduced after 1.466 as far as I tested.
* All `AbortAndRestartOperation` have to do is to add [`NaginatorScheduleAction`](https://github.com/jenkinsci/naginator-plugin/blob/master/src/main/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleAction.java) to the build.

@Jochen-A-Fuerbacher @stefanbrausch
Can you continue this development?
If so, I also want you to write unittests for this new feature.

**Comment by Jochen-A-Fuerbacher** [comment]
We will continue this development, but it could take a little bit time, because we have other projects at the moment with higher priority.

**Comment by ikedam** [comment]
No problem to take a time.
Thanks in advance.
Please let me know if you need any help.

**Comment by Jochen-A-Fuerbacher** [comment]
I implemented naginator integration in build-timeout. But I can't run unit tests of build-timeout plugin. The existing(!) tests fail. According to Maven output, it's a problem with JSP support.
When I force Maven to only test my unit test, it's the same problem. I also figured out that my test fails when calling JenkinsRule.
Do you have any advice?

**Comment by ikedam** [comment]
I don't think I can help you only with that information.

**Comment by Jochen-A-Fuerbacher** [comment]
Ok, I reduced my test to:

`@Rule`
`public BuildTimeOutJenkinsRule j = new BuildTimeOutJenkinsRule();`

`@Test`
`@LocalData`
`public void testAbortAndRestart() throws Exception { }`

where `testAbortAndRestart()` doesn't contain anything.

When I force Maven to only run my "test" with `-Dtest=AbortAndRestartOperationTest`, I get the following error:

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.08 sec <<< FAILURE!
testAbortAndRestart(hudson.plugins.build_timeout.operations.AbortAndRestartOperationTest)  Time elapsed: 2.77 sec  <<< ERROR!
java.lang.NoClassDefFoundError: hudson/model/AbstractBuild$AbstractBuildExecution
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:791)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:423)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:356)
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2451)
	at java.lang.Class.getDeclaredMethods(Class.java:1810)
	at org.jvnet.hudson.annotation_indexer.Index$2$1.fetch(Index.java:101)
	at org.jvnet.hudson.annotation_indexer.Index$2$1.hasNext(Index.java:71)
	at org.jvnet.hudson.annotation_indexer.FilterIterator.fetch(FilterIterator.java:23)
	at org.jvnet.hudson.annotation_indexer.FilterIterator.hasNext(FilterIterator.java:42)
	at hudson.init.InitializerFinder.discoverTasks(InitializerFinder.java:70)
	at hudson.init.InitializerFinder.discoverTasks(InitializerFinder.java:55)
	at org.jvnet.hudson.reactor.TaskBuilder$2.discoverTasks(TaskBuilder.java:40)
	at org.jvnet.hudson.reactor.Reactor.<init>(Reactor.java:128)
	at org.jvnet.hudson.reactor.Reactor.<init>(Reactor.java:133)
	at jenkins.model.Jenkins$6.<init>(Jenkins.java:823)
	at jenkins.model.Jenkins.executeReactor(Jenkins.java:823)
	at jenkins.model.Jenkins.<init>(Jenkins.java:763)
	at hudson.model.Hudson.<init>(Hudson.java:81)
	at org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:486)
	at org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:312)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:431)
	at org.junit.rules.RunRules.evaluate(RunRules.java:18)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
	at org.apache.maven.surefire.junit4.JUnit4TestSet.execute(JUnit4TestSet.java:53)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:123)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:601)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:164)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:110)
	at org.apache.maven.surefire.booter.SurefireStarter.invokeProvider(SurefireStarter.java:172)
	at org.apache.maven.surefire.booter.SurefireStarter.runSuitesInProcessWhenForked(SurefireStarter.java:104)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:70)
Caused by: java.lang.ClassNotFoundException: hudson.model.AbstractBuild$AbstractBuildExecution
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:423)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:356)
	... 53 more

**Comment by ikedam** [comment]
`AbstractBuild.AbstractBuildExecution` isn't available in Jenkins 1.466: https://github.com/jenkinsci/jenkins/commit/14dbcd88d37c675baf6c58682519d129e45cb221
I'm not sure what caused that dependency.
What happens if you removed `LocalData`?

**Comment by ikedam** [comment]
I suppose your current working branch is https://github.com/Jochen-A-Fuerbacher/build-timeout-plugin/tree/naginator : 
* You don't need `WithPlugin` as naginator-plugin will be automatically installed when added to the dependency of pom.xml.
* I encourage not to use `LocalData`, instead, to set up a job inside a test method (using `AbstractProject#getBuildersList` so on).
    * It is easy for other developers to see what is tested.
    * Of course, you can start from `LocalData` and replace it later, or I can replace that after merging your code.

Your trouble might be caused by the mismatch of the Jenkins versions of this plugin and of naginator plugin.
If changing the target version resolves the problem, you can continue the development with that branch.

**Comment by ikedam** [comment]
I found that the problem reproduces also in my environment and I can avoid that in the following way:
```
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>naginator</artifactId>
      <version>1.16-alpha-1</version>
      <optional>true</optional>
      <exclusions>
        <exclusion>
          <groupId>org.jenkins-ci.plugins</groupId>
          <artifactId>matrix-project</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

Details: 
* Matrix-project plugin is split from Jenkins core since 1.561.
* Though naginator plugin itself doesn't work with 1.466, matrix-plugin requires Jenkins >= 1.561.
* Naginator-plugin works with matrix-plugin contained in Jenkins 1.466. And we can resolve the problem by simply excluding matrix-plugin from the dependency tree.

**Comment by Jochen-A-Fuerbacher** [comment]
When I exclude matrix-project like you did, I get the following exception:

`SEVERE: Failed Loading plugin naginator`
`java.io.IOException: Dependency matrix-project (1.2) doesn't exist`

My workaround is, that I changed the required core in pom.xml to 1.609.3. This is the actual LTS version of Jenkins. With that change all existing tests run fine.

**Comment by ikedam** [comment]
> My workaround is, that I changed the required core in pom.xml to 1.609.3. This is the actual LTS version of Jenkins. With that change all existing tests run fine.

Ok. Please continue the development with that change.

I'll make a pull request for naginator-plugin not to depend on matrix-project plugin.
It might be merged in naginator-plugin 1.16.

File: src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        testproject.getBuildWrappersList().add(wrapper);
+        
+        if(System.getProperty("os.name").contains("Windows")){
+            BatchFile builder = new BatchFile("ping -n 500 127.0.0.1 &amp;gt;nul");
+            testproject.getBuildersList().add(builder);
+        }else{
+            Shell builder = new Shell("sleep 500");
+            testproject.getBuildersList().add(builder);
+        }
+
```
Comment (ikedam) [inline src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java]: You can use [`SleepBuilder`](http://javadoc.jenkins-ci.org/org/jvnet/hudson/test/SleepBuilder.html) instead.
https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/test/java/hudson/plugins/build_timeout/operations/BuildStepOperationTest.java#L148

File: src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        assertEquals(Result.ABORTED, testproject.getFirstBuild().getResult());
+        assertEquals(Result.ABORTED, testproject.getLastBuild().getResult());
+    }
+    
+    @Test
+    public void testAbortAndRestartTwice() throws Exception {
+        
+        FreeStyleProject testproject = j.createFreeStyleProject();
+
+        AbsoluteTimeOutStrategy strategy = new AbsoluteTimeOutStrategy(Integer.toString(3)); //timeoutMinutes
```
Comment (ikedam) [inline src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java]: You can save the test execution time by using [`QuickTimeoutStrategy`](https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/test/java/hudson/plugins/build_timeout/QuickBuildTimeOutStrategy.java) (available only in tests).
https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/test/java/hudson/plugins/build_timeout/operations/BuildStepOperationTest.java#L142

File: src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        
+        assertTrue(testproject.getBuilds().size()==0);
+                
+        
+        testproject.scheduleBuild(new Cause.UserIdCause());
+        try{
+            Thread.sleep((3*60*1000)+20000);
+        }catch(InterruptedException e){
+            //Nothing todo here.
+        }
```
Comment (ikedam) [inline src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java]: You can use [`JenkinsRule#waitUntilNoActivity`](http://javadoc.jenkins-ci.org/org/jvnet/hudson/test/JenkinsRule.html#waitUntilNoActivity()) instead.
https://github.com/jenkinsci/naginator-plugin/blob/naginator-1.16-alpha-1/src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java#L176

It blocks until all builds (including rescheduled builds) finish.

**Comment by ikedam** [comment]
Though this work may be still in progress, I added some comments that make it easy for you to write and run unit tests.
Please let me know if this request is ready for full reviews.

**Comment by ikedam** [comment]
You can ignore errors from findbugs if raised in lines you didn't change.
Newer Jenkins core are fully annotated with findbugs and the check gets more strict when upgrading the target version.
Of course, you can fix them if you like it.

**Comment by Jochen-A-Fuerbacher** [comment]
I changed the code you commented.
You can do a full review of this request.

Thank you!

File: src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        
+        
+        testproject.getBuildersList().add(new SleepBuilder(5*60*1000)); //5 minutes
+                       
+        
+        testproject.scheduleBuild(new Cause.UserIdCause());
+
+        j.waitUntilNoActivityUpTo(25000);
+            
+        assertTrue(testproject.getFirstBuild() != null);
```
Comment (ikedam) [inline src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java]: Let's use `assertNotNull`. It helps you much when the test fails unexpectedly as it displays the actual unexpected value in the log.

File: src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Test
+    public void testAbortAndRestartOnce() throws Exception {
+        
+        FreeStyleProject testproject = j.createFreeStyleProject();
+
+        QuickBuildTimeOutStrategy strategy = new QuickBuildTimeOutStrategy(5000);
+        AbortAndRestartOperation operation = new AbortAndRestartOperation(1); //Number of restarts
+        LinkedList<BuildTimeOutOperation> list = new LinkedList<BuildTimeOutOperation>();
+        list.add(operation);
```
Comment (ikedam) [inline src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java]: JFYI: [`Arrays.asList`](http://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html#asList(T...)) is useful for this purpose.

File: src/main/resources/hudson/plugins/build_timeout/operations/AbortAndRestartOperation/help.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<div>
+Count of maximum restarts. 0 means without a limit.</a>.
+</div>
```
Comment (ikedam) [inline src/main/resources/hudson/plugins/build_timeout/operations/AbortAndRestartOperation/help.html]: This file should be `help-maxRestarts.html` (the help text for `maxRestarts`).

File: src/main/resources/hudson/plugins/build_timeout/operations/AbortAndRestartOperation/help-description.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<div>
+Maximum of restarts.
+</div>
```
Comment (ikedam) [inline src/main/resources/hudson/plugins/build_timeout/operations/AbortAndRestartOperation/help-description.html]: This file isn't used.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            build.addAction(new com.chikli.hudson.plugin.naginator.NaginatorScheduleAction(this.maxRestarts, sd, false));
+        }
+        return true;
+    }
+        
+    @Extension 
+    public static class DescriptorImpl extends BuildTimeOutOperationDescriptor {
+        @Override
+        public String getDisplayName() {
+            return "Abort and restart the build";
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: Please prepare for the internationalization with Messages.properties.
Example:
* https://github.com/jenkinsci/build-timeout-plugin/blob/master/src/main/resources/hudson/plugins/build_timeout/operations/Messages.properties
* https://github.com/jenkinsci/build-timeout-plugin/blob/master/src/main/java/hudson/plugins/build_timeout/operations/AbortOperation.java#L75

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        
+        Executor e = build.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.ABORTED);
+        }
+        
+        if(isPresent()){
+            FixedDelay sd = new FixedDelay(0); //Reschedule now!
+            build.addAction(new com.chikli.hudson.plugin.naginator.NaginatorScheduleAction(this.maxRestarts, sd, false));
+        }
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: You can have this operation displayed only when naginator-plugin is installed with overriding `BuildTimeOutOperationDescriptor#isApplicable`
https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/main/java/hudson/plugins/build_timeout/BuildTimeOutOperationDescriptor.java#L49

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    @DataBoundConstructor
+    public AbortAndRestartOperation(int maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+    
+    private static boolean isPresent() {
+        try {
+            Class.forName("com.chikli.hudson.plugin.naginator.NaginatorScheduleAction");
+            return true;
+        } catch (Throwable ex) {
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: Should be `ClassNotFoundException`.
Catching all exceptions and disposing them may miss unexpected exceptions.

File: src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java [hunk 14]
```diff
@@ -20,7 +20,7 @@
 /**
  * If the build reaches <tt>deadlineTime</tt>, it will be terminated.
  * 
- * @author Fernando Migu�lez Palomo <fernando.miguelez@gmail.com>
+ * @author Fernando Miguelez Palomo <fernando.miguelez@gmail.com>
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java]: Please do not change an unrelated file.

**Comment by ikedam** [comment]
It's better to accept variable expressions for `maxRestarts`.
Really sorry for a late request, but it will get hard to change a type of a field after making a release.
You can expand variable with `build.getEnvironment(listener).expand(string);`.
Would you try that?

File: src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java (hunk 14 above)
Comment (Jochen-A-Fuerbacher) [inline src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java]: With the é in the name, I get encoding error when building. This happens since I changed the required jenkins core verion to 1.609.3. So I changed the é to e.

**Comment by Jochen-A-Fuerbacher** [comment]
I changed `maxRestarts` to accept variable expressions.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    @Extension 
+    public static class DescriptorImpl extends BuildTimeOutOperationDescriptor {
+        @Override
+        public String getDisplayName() {
+            return Messages.AbortAndRestartOperation_DisplayName();
+        }
+        
+        @Override
+        public boolean isApplicable(Class<? extends AbstractProject<?,?>> jobType) {
+            return true;
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: `isPresent()` is expected to be used here.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        
+        if(isPresent()){
+            FixedDelay sd = new FixedDelay(0); //Reschedule now!
+            int maxRestarts = 0;
+            try {
+                maxRestarts = Integer.parseInt(build.getEnvironment(listener).expand(this.maxRestarts));
+                build.addAction(new com.chikli.hudson.plugin.naginator.NaginatorScheduleAction(maxRestarts, sd, false));
+            } catch (IOException e1) {
+                log.log(Level.WARNING, "Failed to expand environment variables. ", e1);
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: It's useful for users to output logs to the console log of the build.
`listener.getLogger()` provides `PrintStream` for that.

**Comment by ikedam** [comment]
> With the é in the name, I get encoding error when building.

Hmm...I'll have a look on that. I think I can resolve that by adding encoding instructions in pom.xml.

**Comment by ikedam** [comment]
Thanks for the update.
I added a few more comments, but I think they can be fixed easily.

Please wait for a little more time till naginator 1.16 is released.
I believe we can revert the target version after that and it makes many things easier (e.g. firebug no longer claims errors).

File: src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java [hunk 17]
```diff
@@ -20,7 +20,7 @@
 /**
  * If the build reaches <tt>deadlineTime</tt>, it will be terminated.
  * 
- * @author Fernando Migu�lez Palomo <fernando.miguelez@gmail.com>
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/impl/DeadlineTimeOutStrategy.java]: I found that this was not UTF-8. This should be 0xc3a9 in UTF-8, but was 0xe9 (supposed latin-1).

**Comment by Jochen-A-Fuerbacher** [comment]
Thank you for your comments. I changed the code you commented.
Anything I should do because of the encoding problem?

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    public AbortAndRestartOperation(String maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+    
+    private static boolean isPresent() {
+        try {
+            Class.forName("com.chikli.hudson.plugin.naginator.NaginatorScheduleAction");
+            return true;
+        } catch (ClassNotFoundException ex) {
+            log.log(Level.WARNING, "Naginator not available. ", ex);
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: This message will be output too often called from `isApplicable` (every time a user opens project configuration page).
Please change this to FINEST (output only for debugging purpose) and output warnings to console logs of builds instead (that is, output inside `perform`).

**Comment by ikedam** [comment]
> Anything I should do because of the encoding problem?

Nothing to do for now.
We'll revert the target Jenkins version when the new version of naginator plugin is released, and we can revert that file at that time.

I'll change the encoding of that file in another commit (It's not a good idea to do that in this request).

**Comment by ikedam** [comment]
Sorry for having you wait for so long time!
Naginator-plugin 1.16 was released. It may take some time till it's available in the update center, but should be already available in the maven repository.

Would you do followings?:
* Change the version of naginator plugin (1.16), and revert the version of Jenkins core (1.466).
    * This should resolve errors in findbugs.
* Revert the change in DeadlineTimeOutStrategy.java.
    * You can remove the change from the commit history using git features (see below).
    * I'll fix this problem in another ticket.
* Squash commits.
    * There are many commits without comments.
    * You can use this by using `git rebase` and overwrite the pull request with `git push --force`.
    * For details, please refer https://help.github.com/articles/about-git-rebase/
    * If you don't Interested in Git, I can do that instead. But I'd like you to do that as
        * This pull request won't mark as "merged" if I do that instead.
        * Git is powerful tool, and I believe above commands will help your future developments.

**Comment by Jochen-A-Fuerbacher** [comment]
Hello,

in the next few weeks I don't have the time to work on that plugin. But it's in my backlog and I'm looking forward to work on this as soon as possible.

**Comment by ikedam** [comment]
Rebased and recreated #54.

**Comment by ikedam** [comment]
Released in build-timeout 1.17.
